demoapps/passwordlogin/java/org/openyolo/demoapp/passwordlogin/LoginViewModel.java

```diff
@@ -0,0 +1,513 @@
+import android.databinding.ObservableField;
+import android.net.Uri;
+import android.support.annotation.AnyThread;
+import android.support.annotation.Keep;
+import android.support.annotation.MainThread;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.support.annotation.StringRes;
+import android.support.annotation.WorkerThread;
+import android.util.Log;
+import android.util.Patterns;
+import android.view.View;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.openyolo.api.CredentialClient;
+import org.openyolo.demoapp.passwordlogin.userdata.HashUtil;
+import org.openyolo.demoapp.passwordlogin.userdata.UserDataSource;
+import org.openyolo.protocol.AuthenticationDomain;
+import org.openyolo.protocol.AuthenticationMethods;
+import org.openyolo.protocol.Credential;
+import org.openyolo.protocol.CredentialRetrieveRequest;
+import org.openyolo.protocol.CredentialRetrieveResult;
+import org.openyolo.protocol.CredentialSaveRequest;
+import org.openyolo.protocol.CredentialSaveResult;
+import org.openyolo.protocol.Hint;
+import org.openyolo.protocol.HintRetrieveRequest;
+import org.openyolo.protocol.HintRetrieveResult;
+import org.openyolo.protocol.RetrieveBbqResponse;
+
+/**
+ * The view model for {@link LoginActivity}. Defines all the logic for authenticating a user
+ * which an email address and password, including interacting with any available OpenYOLO provider
+ * to directly retrieve or generate credentials.
+ */
+@WorkerThread
+public final class LoginViewModel extends ObservableViewModel {
+
+    private static final String TAG = "LoginViewModel";
+
+    private static final int MIN_PASSWORD_LENGTH = 4;
+
+    /**
+     * Indicates whether a progress bar should be displayed to the user, while an asynchronous
+     * request is made.
+     */
+    public final ObservableBoolean loading = new ObservableBoolean(true);
+
+    /**
+     * Describes the asynchronous request that is occurring to the user.
+     */
+    public final ObservableField<String> loadingLabel = new ObservableField<>("Loading…");
+
+    /**
+     * The prompt that is displayed to the user for manual authentication, instructing them what
+     * to do next.
+     */
+    public final ObservableField<String> authPrompt = new ObservableField<>("");
+
+    /**
+     * The email address the user has entered, or that was retrieved from their OpenYOLO provider.
+     */
+    public final ObservableField<String> email = new ObservableField<>("");
+
+    /**
+     * The error string displayed below the email text input element. If the string is empty,
+     * the error is not displayed.
+     */
+    public final ObservableField<String> emailError = new ObservableField<>("");
+
+    /**
+     * Whether the password entry field should be displayed at this time.
+     */
+    public final ObservableBoolean showPasswordField = new ObservableBoolean(false);
```

Comment:
Nit: rename to shouldShowpasswordField

Reply:
Leaving as showPasswordField for consistency with showLoading.